webapp/graphite/tags/redis.py

```diff
@@ -27,68 +29,119 @@ def __init__(self):
-          pattern = re.compile(spec)
-          matches_empty = not pattern.match('')
-
-          values = [value for value in self.r.sscan_iter('tags:' + tag + ':values') if pattern.match(value) is None]
-
-        else:
-          raise ValueError("Invalid operator %s" % operator)
-
-        # if we're matching the empty value, union with the set of series that don't have the tag
-        if matches_empty:
-          # store series that don't have this tag in a temp key
-          pipe.sdiffstore('temp:' + tag + ':empty', 'series', 'tags:' + tag + ':series')
-
-          # store union of series without this tag and all series that match the spec in a temp key
-          pipe.sunionstore('temp:' + tagspec, 1 + len(values), 'temp:' + tag + ':empty', *['tags:' + tag + ':values:' + value for value in values])
-        # otherwise only return series for matching values
-        else:
-          # if this query matched no values, just short-circuit since the result of the final intersect will be empty
-          if not values:
-            return []
-
-          # store union of all series that match the spec in a temp key
-          pipe.sunionstore('temp:' + tagspec, len(values), *['tags:' + tag + ':values:' + value for value in values])
-
-        all_match_empty = all_match_empty and matches_empty
-
-      if all_match_empty:
-        raise ValueError("At least one tagspec must not match the empty string")
-
-      # get the final result from the intersection of the temp lists we created above
-      pipe.sinter(*['temp:' + tagspec for tagspec in tags])
-
-      results = pipe.execute()
-
-    return sorted(list(results[-1]))
+    selector = None
+    selector_cnt = None
+    filters = []
+
+    # loop through tagspecs, look for best = or =~ match
+    for tagspec in tags:
+      (tag, operator, spec) = self.parse_tagspec(tagspec)
+
+      if operator == '=':
+        matches_empty = spec == ''
+        if not matches_empty and (not selector or selector[1] != '='):
+          cnt = self.r.scard('tags:' + tag + ':values:' + spec)
+          if not selector or selector_cnt > cnt:
+            if selector:
+              filters.append(selector)
+            selector = (tag, operator, spec)
+            selector_cnt = cnt
+            continue
+        filters.append((tag, operator, spec))
+
+      elif operator == '=~':
+        pattern = re.compile(spec)
+        matches_empty = bool(pattern.match(''))
+        if not matches_empty and (not selector or selector[1] != '='):
+          cnt = self.r.scard('tags:' + tag + ':values')
+          if not selector or selector_cnt > cnt:
+            if selector:
+              filters.append(selector)
+            selector = (tag, operator, pattern)
+            selector_cnt = cnt
+            continue
+        filters.append((tag, operator, pattern))
+
+      elif operator == '!=':
+        matches_empty = spec != ''
+        if not matches_empty and (not selector or selector[1] != '='):
```

Comment:
i thought we're requiring every query to contain at least one expression with the `=` or `=~` operator.

Reply:
Technically `tag!=` is a valid query, because it doesn't match the empty string, it's the same thing as `tag=~.+`